src/System.Net.Sockets/tests/FunctionalTests/KeepAliveTest.netcoreapp.cs

```diff
@@ -0,0 +1,98 @@
+            }
+        }
+
+        private bool RunningOnWindowsBelow10v1709
+        {
+            get
+            {
+                if (!RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+                    return false;
+                Version version = Environment.OSVersion.Version;
+                return version.Major < 10 || version.Major == 10 && version.Build < 16299;
+            }
+        }
+
+        [Fact]
+        public void Socket_KeepAlive_Disabled_By_Default()
+        {
+            using (Socket socket = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp))
+            {
+                Assert.Equal<int>(0, (int)socket.GetSocketOption(SocketOptionLevel.Socket, SocketOptionName.KeepAlive));
+            }
+        }
+
+        [Fact]
+        public void Socket_KeepAlive_Enable()
+        {
+            using (Socket socket = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp))
+            {
+                socket.SetSocketOption(SocketOptionLevel.Socket, SocketOptionName.KeepAlive, true);
+                Assert.NotEqual<int>(0, (int)socket.GetSocketOption(SocketOptionLevel.Socket, SocketOptionName.KeepAlive));
+            }
+        }
+
+        [Fact]
+        public void Socket_KeepAliveState_Set_RetryCount()
+        {
+            // TcpKeepAliveRetryCount can be managed via *SocketOption starting from Windows 10 version 1703
+            if (RunningOnWindowsBelow10v1703)
+                return;
+
+            const int retryCount = 60;
+            using (Socket socket = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp))
+            {
+                socket.SetSocketOption(SocketOptionLevel.Tcp, SocketOptionName.TcpKeepAliveRetryCount, retryCount);
+                Assert.Equal<int>(retryCount, (int)socket.GetSocketOption(SocketOptionLevel.Tcp, SocketOptionName.TcpKeepAliveRetryCount));
+            }            
+        }
+
+        [Fact]
+        public void Socket_KeepAliveState_Set_Time()
+        {
+            // TcpKeepAliveTime can be managed via *SocketOption starting from Windows 10 version 1709
+            if (RunningOnWindowsBelow10v1709)
+                return;
+
+            const int time = 5;
+            using (Socket socket = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp))
+            {
+                socket.SetSocketOption(SocketOptionLevel.Tcp, SocketOptionName.TcpKeepAliveTime, time);
+                Assert.Equal<int>(time, (int)socket.GetSocketOption(SocketOptionLevel.Tcp, SocketOptionName.TcpKeepAliveTime));
+            }            
+        }
+
+        [Fact]
+        public void Socket_KeepAliveState_Set_Interval()
+        {
+            // TcpKeepAliveInterval can be managed via *SocketOption starting from Windows 10 version 1709
+            if (RunningOnWindowsBelow10v1709)
+                return;
```

Comment:
What happens on earlier versions of Windows?  It'd be good to have some negative tests that validate whatever exception we expect to get if the option setting doesn't work.

Reply:
They should throw a SocketException, but do you currently run tests on different Windows versions?

TCP_KEEPIDLE [should](https://msdn.microsoft.com/en-us/library/windows/desktop/ms738596.aspx) be available starting from 1709 but on a 1703 machine it does not throw any exception 😲

Reply:
> do you currently run tests on different Windows versions?

Yes.  We have a larger matrix separate from PRs, but even for PRs, every change runs against Windows 7, Windows 8.1, and Windows 10.